components/net_traits/response.rs

```diff
@@ -225,24 +225,50 @@ impl Response {
         response
     }
 
-    pub fn metadata(&self) -> Result<Metadata, NetworkError> {
+    pub fn metadata(&self) -> Result<FetchMetadata, NetworkError> {
+        fn init_metadata(metadata: &mut Metadata, response: &Response) {
+            metadata.set_content_type(match response.headers.get() {
+                Some(&ContentType(ref mime)) => Some(mime),
+                None => None
+            });
+            metadata.headers = Some(Serde(response.headers.clone()));
+            metadata.status = response.raw_status.clone();
+            metadata.https_state = response.https_state;
+        };
+
         let mut metadata = if let Some(ref url) = self.url {
-            Metadata::default(url.clone())
+            Some(Metadata::default(url.clone()))
+        } else if self.is_network_error() {
+            return Err(NetworkError::Internal("Cannot extract metadata from network error".to_owned()));
         } else {
-            return Err(NetworkError::Internal("No url found in response".to_string()));
+            None
         };
 
-        if self.is_network_error() {
-            return Err(NetworkError::Internal("Cannot extract metadata from network error".to_string()));
+        if let Some(ref mut m) = metadata {
+            init_metadata(m, self);
         }
 
-        metadata.set_content_type(match self.headers.get() {
-            Some(&ContentType(ref mime)) => Some(mime),
-            None => None
-        });
-        metadata.headers = Some(Serde(self.headers.clone()));
-        metadata.status = self.raw_status.clone();
-        metadata.https_state = self.https_state;
-        return Ok(metadata);
+        if let Some(ref response) = self.internal_response {
+            match response.url {
+                Some(ref url) => {
+                    let mut unsafe_metadata = Metadata::default(url.clone());
+                    init_metadata(&mut unsafe_metadata, response);
+
+                    Ok(FetchMetadata::Filtered {
+                        filtered: match metadata {
+                            Some(m) => FilteredMetadata::Transparent(m),
+                            None => FilteredMetadata::Opaque
+                        },
+                        unsafe_: unsafe_metadata
+                    })
+                }
+                None => Err(NetworkError::Internal("No url found in unsafe response".to_owned()))
+            }
+        } else {
+            match metadata {
+                Some(m) => Ok(FetchMetadata::Unfiltered(m)),
+                None => Err(NetworkError::Internal("No url found in response".to_owned()))
```

Comment:
I would rather make this an `unreachable!()`, because we shouldn't be able to have a URL-less response with no internal response.